File: 03_Data_Structures_&_Counting/2108_Statistics/Answer2108.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.roundToInt

fun main(){
  val br = BufferedReader(InputStreamReader(System.`in`))
  val n = br.readLine().toInt()
  val nums = IntArray(n)
  for(i in 0 until n) nums[i] = br.readLine().toInt()
  nums.sort()

  val countsObj = HashMap<Int,Int>()

  for(i in 0 until n){
    val it = nums[i]
    if(countsObj[it] != null){
      countsObj[it] = countsObj[it]!! + 1
    } else {
      countsObj[it] = 1
    }
  }

  val countsArr = mutableListOf<IntArray>()
  for(i in countsObj.keys){
    countsObj[i]?.let{countsArr.add(element = intArrayOf(i, countsObj[i]!!))}
  }
  countsArr.sortWith{a, b ->
    if(a[1] == b[1]){
      a[0] - b[0]
    } else {
      b[1] - a[1]
    }
  }
  val ave = nums.average()
  val roundAve = ave.roundToInt()

  var mode = 0
  if(countsArr.size > 1) {
    mode = if(countsArr[0][1] == countsArr[1][1]) countsArr[1][0] else countsArr[0][0]
  } else mode = countsArr[0][0]

  val resultStr = StringBuilder()
  resultStr.append("${roundAve}\n")
  resultStr.append("${nums[n / 2]}\n")
  resultStr.append("${mode}\n")
  resultStr.append("${nums[n - 1] - nums[0]}")
  print(resultStr)
}

File: 03_Data_Structures_&_Counting/2108_Statistics/Answer2108.rs
use std::io::{stdin, Read};
use std::collections::HashMap;
use std::fmt::Write;

fn main(){
  let mut s = String::new();
  stdin().read_to_string(&mut s).unwrap();
  let inputs:Vec<i32>= s.split_whitespace().map(|x| x.parse().unwrap()).collect();
  let n:usize = inputs[0] as usize;
  let mut nums:Vec<i32> = inputs[1..].to_vec();
  nums.sort();

  let mut counts_obj:HashMap<i32, i32> = HashMap::new();

  for it in &nums {
    if counts_obj.contains_key(&it) {
      *counts_obj.get_mut(&it).unwrap() += 1;
    } else {
      counts_obj.insert(*it, 1);
    }
  }

  let mut counts_arr:Vec<[i32; 2]> = Vec::new();
  for (key, value) in &counts_obj {
    counts_arr.push([*key, *value]);
  }
  counts_arr.sort_by(|a, b| {
    if a[1] == b[1] {
      a[0].cmp(&b[0])
    } else {
      b[1].cmp(&a[1])
    }
  });
  let sum:i32 = nums.iter().sum();
  let ave:f64 = sum as f64 / n as f64;
  let round_ave = ave.round() as i32;

  let mode;
  if counts_arr.len() > 1 {
    mode = if counts_arr[0][1] == counts_arr[1][1] {counts_arr[1][0]} else {counts_arr[0][0]}
  } else {mode = counts_arr[0][0];}

  let mut result_str = String::new();
  write!(result_str,"{}\n{}\n{}\n{}", round_ave, nums[n/2], mode, nums[n-1] - nums[0]).unwrap();
  print!("{}", result_str);
}